### DISCUSSION TITLE
How to match PyMuPdf output to pdf2image output?

### DISCUSSION THREAD

**Discussion by arun-rangarajan** [pymupdf/PyMuPDF#913] [Looking for help] [answered]
Looking for help on [this question](https://stackoverflow.com/questions/66284722/how-to-match-pymupdf-output-to-pdf2image-output) asked on Stackoverflow.

---
Got a PDF to PNG converter built with pdf2image. It is quite slow for converting PDF to PNG images. For example, a 7-page PDF document takes 10 seconds to get split into PNG images even with thread_count set to 4 on a 4-core machine (Standard B4ms Azure VM).

Tried PyMuPdf and it ran much faster (only 800 ms) with default scaling:
`mat = fitz.Matrix(1, 1)`

Then I realized that the PNGs that are output with default scaling are much smaller compared to those of pdf2image, so increased the scaling to match the PNGs output by pdf2image. I had to use a scale factor of 2.7777 for the pixel sizes to match up with pdf2image, so
`mat = fitz.Matrix(2.7777, 2.7777)`

This took 3 seconds to run, but still much faster compared to pdf2image.

The images output by PyMuPdf looked quite identical to those of pdf2image to my eyes, but they actually differ. Our downstream processing (an object detection model) which uses these PNG's also produces different results.

Looking at pdf2image doc, we have just used the default dpi of 200. How does one translate this setting to PyMuPdf to get the exact same output? I tried [setResolution](https://pymupdf.readthedocs.io/en/latest/pixmap.html#Pixmap.setResolution) of 200, but that didn't help.

---

**Comment by JorjMcKie** [comment, marked as answer]
>  I tried setResolution of 200, but that didn't help.

But the resolution was correctly set to the value, wasn't it? So, what **_was_** the difference?
Anyway, PyMuPDF also supports using Pillow for pixmap output, try this `pix.pillowWrite("%02i.png" % page.number, dpi=(200, 200))`.
The parameters of `pillowWrite()` are passed through to Pillow's `Image.save()` method unchanged. This should enable you to make the output as equal as desired.

**Reply by arun-rangarajan** [reply]
Thx for your answer (and the project), @JorjMcKie. 

This is the pdf2image code I had:

```
im = pdf2image.convert_from_path(file_path, fmt="png", thread_count=8)
for i, image in enumerate(im, start=1):
    image.save(f"page-{i}.png")
```

This outputs PNG's that are 1700 x 2200 px for [this PDF](https://github.com/pymupdf/PyMuPDF/files/6013776/tracemonkey.pdf).

And here was my first attempt at the equivalent PyMuPdf code:

```
doc = fitz.open(file_path)
zoom = 1 # change resolution instead of zoom
mat = fitz.Matrix(zoom, zoom)
for i, page in enumerate(doc, start=1):
    pix = page.getPixmap(matrix=mat, alpha=False)
    pix.setResolution(200, 200)
    pix.writePNG(f"page-{i}.png")
```

This outputs 612 x 792 px.

After your suggestion, I tried:

```
doc = fitz.open(file_path)
zoom = 1 # change dpi instead of zoom
mat = fitz.Matrix(zoom, zoom)
for i, page in enumerate(doc, start=1):
    pix = page.getPixmap(matrix=mat, alpha=False)
    pix.pillowWrite(f"page-{i}.png", dpi=(200, 200))
```

This also outputs 612 x 792 px.

Only by changing `zoom` above, the pixel sizes could be changed. I don't think I understand the relationship between `zoom`, and dpi/resolution.

**Reply by JorjMcKie** [reply]
The zoom parameter controls how many pixels are created per each **point** of the page.
For a letter page with 612 x 792 points (which is - by using 72 points per 1 inch - 8.5 x 11.0 inches), the resulting pixmap also has this dimension if you zoom = 1.
By assigning a dpi value, you control the **printed** page size. MuPDF's default of 96 dpi therefore leads to a print page size of 6.38 x 8.25 inches.
So, to get a print page size of letter at 96 dpi, must use zoom = 96 / 72 = 1.3333, and similarly for getting a letter print size at 200 x 200 dpi, use zoom = 200 / 72 = 2,7777.

**Reply by JorjMcKie** [reply]
But independently from zoom, also use dpi to determine the print page size.

**Reply by arun-rangarajan** [reply]
Thx again, @JorjMcKie. 

I used zoom of (200 / 72) = 2.7777 and set dpi like this:
`pix.pillowWrite(f"page-{i}.png", dpi=(200, 200))`

Now the output PNG's are 1700 x 2200 px.

However pdf2image file sizes are 10% to 20% larger compared to pymupdf's. Not sure if it's possible to make the outputs identical.

**Reply by JorjMcKie** [reply]
I looked up pdf2image - they seem to use pillow as well. Obviously with a different PNG compression parameter.
Maybe consult Pillow docu, they may have a way to set PNG compression ratio. Internally, PNG use gzip routines to compress image data, and for this there exist the usual ratios 0 through 9, with a default of 6 (supposedly a good balance between speed and file size).

**Reply by arun-rangarajan** [reply]
Thx @JorjMcKie. Will investigate this further.
